Add GitHub Actions to eventually replace AppVeyor

RevoluPowered opened the pull request:
This is based on @qarmin's work in #38433

It *only* builds windows at the moment and provides:
- windows editor binaries which you can download.
- pull request building
- **tested and working caching**
- explicit python versioning 
- faster builds

It's really meant to just replace AppVeyor right now in a smaller step so Godot can evaluate it internally.

**Example/overview of the actions**
- https://github.com/RevoluPowered/godot/actions

**Windows binaries example**
https://github.com/RevoluPowered/godot/actions/runs/165602609

**MSVC caching working with scons 4.0.0**
https://github.com/RevoluPowered/godot/runs/860871691?check_suite_focus=true

I customized the cache plugin to fix a bug in the tool on windows, basically, it was breaking the extraction of the cache on the builds. I reported upstream to actions/cache and applied the fix to this file, so it will just work.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+      branches: [ master, 3.2, 3.1, 3.0 ]
+env:
+  SCONS_CACHE_MSVC_CONFIG: true
+  SCONS_CACHE_LIMIT: 2048
```
For the reference, do you know what's the max cache size for GH actions? Is the cache max size split across all jobs, or is it a max per job?

RevoluPowered replied:
Yes, the cache size is limited per repository to 5GB. It will then throw out the cache, by 'oldest first'.

Godot builds roughly a 400-500 MB cache on windows, so this cache for release and editor is approximately 1GB for the windows platform.

Artifacts are counted separately.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+    # You can test your matrix by printing the current Python version
+    - name: Display Python version
+      run: python -c "import sys; print(sys.version)"
+    - name: Installing scons and pywin32
```
Does it still make a difference to use pywin32? AFAIK we no longer document the need to install it (CC @Calinou).

Calinou replied:
As far as I know, it shouldn't be required anymore with recent Python versions, even if SCons warns you about it.

RevoluPowered replied:
It's still requested, so someone should look at this

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+  windows-editor:
+    runs-on: "windows-latest"
+    name: Windows Editor (release_debug)
+    steps:
```
There's a lot of steps, do we need to be so fine-grained when the output we're most interested in will be: What python/scons env we have, and the actual Godot build logs and upload?

I'd suggest grouping all the Python setup under one named step.

RevoluPowered replied:
I smushed the python together so this should be better, but I would like to keep the python installer separate since it is official code from their plugin basically means updating this in the future if they change it is easier

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+    steps:
+    - uses: actions/checkout@v2 
+    # cache must be put in the github workspace directory on windows to be persisted correctly
+    - name: "force cache path to be correct"
```
Some steps start with capital letters, others with lowercase. We should be consistent (probably capital letter everywhere).

RevoluPowered replied:
I have fixed this in the code it should now be consistent

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+        scons --version
+    - name: Compilation
+      run: |
+        scons -j2 verbose=yes warnings=all werror=yes platform=windows target=release_debug
```
Is it confirmed that GitHub Actions has 2 cores, and only 2?

RevoluPowered replied:
Yes only 2 cores per machine, you only get more if you pay, or if you self host.

side topic: Self hosting is fairly easy and has been tested already it works, with very minor tweaks. Requires a dedicated vm which is secured to work for the godot engine though, so would recommend we stay with their stuff until @hpvb can look at setting something up possibly?

I tested with windows 10 already and the self hosting feature just works with no specific or non sane options, one minor edit to the permissions on the vm and it works. To be clear this config explictly will only build using the Github Action servers right now.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+      with:
+        name: windows-editor (x64)
+        path: bin/godot.windows.opt.tools.64.exe
+  windows-runtime:
```
We use "template" for this so it's better to stay consistent (same for the job name and for the uploaded artifact's name).

RevoluPowered replied:
I can't really change the name of the file itself, it will be whatever it's called by scons, I can, however, edit the name to be 'windows-template', is this what you mean?

RevoluPowered replied:
miread the line number, yeah so this is easy to edit

RevoluPowered replied:
I have fixed this here and commented

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+    - name: Compilation
+      run: |
+        scons -j2 verbose=yes warnings=all werror=yes platform=windows target=release tools=no
+    - name: publishing godot windows-editor
```
That's the template, not editor.

RevoluPowered replied:
copy pasted sorry will fix this now

RevoluPowered replied:
This has been fixed :) and commented out as discussed.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+    - name: publishing godot windows-editor
+      uses: actions/upload-artifact@v1
+      with:
+        name: windows-runtime-release (x64)
```
Is this the name of the resulting download? If so, can we make it more unique by including the commit hash (shortened to 7 or 9 chars)?

RevoluPowered replied:
yeah i can do this

RevoluPowered replied:
Corrected and disabled as discussed in IRC

akien-mga commented:
I suggest squashing the two commits together, and you can credit @qarmin as co-author in the commit log:
```
Co-authored-by: Rafał Mikrut <mikrutrafal54@gmail.com>
```

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,94 @@
+    pull_request:
+    push:
+      branches: [ master, 3.2, 3.1, 3.0 ]
+env:
```
Add some vertical space (one line) before top level sections to make things more readable.

As in the official template:
```
# This is a basic workflow to help you get started with Actions

name: CI

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the master branch
on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
    # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
    - uses: actions/checkout@v2

    # Runs a single command using the runners shell
    - name: Run a one-line script
      run: echo Hello, world!

    # Runs a set of commands using the runners shell
    - name: Run a multi-line script
      run: |
        echo Add other actions to build,
        echo test, and deploy your project.
```

RevoluPowered replied:
Sure I will add some comments on this too to explain what each step is

RevoluPowered replied:
Done!

RevoluPowered commented:
I still have to check the caching works with the edits requested so will need to wait

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,121 @@
+        scons --version
+    - name: Compilation
+      run: |
+        scons -j2 compdb=no verbose=yes warnings=all werror=yes platform=windows target=release tools=no
```
What does `compdb=no` do? We don't have this option currently.

RevoluPowered replied:
sorry meant compiledb=no :)

RevoluPowered replied:
should turn off compilation database

RevoluPowered replied:
removed entirely

RevoluPowered committed:
Use GitHub Actions instead of AppVeyor for Windows platform

This was based on quarmin's initial configuration, we had compile issues with AppVeyor being super slow and GitHub actions will take less time and also manage a full rebuild in 30 minutes.

This adds cache handling and will work with MSVC and scons 4.0, it will build for every PR submitted to the Godot Engine, and also for the branches specified. I have tested the caching and it seems to be working.

I left the 'publish artefacts' disabled until we can request more storage from Microsoft, 5 GB is far to low for us and we would eat this limit very fast. (it is tested and works fine)

Co-authored-by: Rafał Mikrut <mikrutrafal54@gmail.com>

akien-mga commented:
Thanks a ton! Amazing work :)

akien-mga commented:
Cherry-picked for 3.2.3.

bruvzg commented:
Seems like GitHub Actions have some issues with the disk space:
https://github.com/godotengine/godot/pull/38900/checks?check_run_id=872790339

```
LINK : fatal error LNK1201: error writing to program database 'D:\a\godot\godot\bin\godot.windows.opt.tools.64.pdb'; check for insufficient disk space, invalid path, or insufficient privilege
```

akien-mga commented:
Seems to be this issue: https://github.community/t/bigger-github-hosted-runners-disk-space/17267/9

No resolution in sight yet so I guess we need to tweak build options to make a leaner build... maybe we can disable debug symbols?

bruvzg commented:
> it seems Github does not even set a static limit set, it’s 10 GB “Guaranteed”

What's taking that much space? My local Godot folder with macOS Universal, MinGW and Linux builds all done in the same source tree and with multiple custom modules added is only 10.5 GB large.

akien-mga commented:
Maybe the cache? It's configured to be max 8 GB, and it should be less in theory, but once uncompressed + adding the debug build symbols, it might reach 10 GB.

aaronfranke commented:
Why was this cherry-picked for 3.2? Isn't changing the CI system a major change that can break things, and a change that is not at all necessary for a stable branch? Shouldn't this also be changed for 3.1 and 3.0 if this is changed for 3.2?

akien-mga commented:
> Why was this cherry-picked for 3.2? Isn't changing the CI system a major change that can break things, and a change that is not at all necessary for a stable branch?

Well CI doesn't break anything, it has no incidence on the compiled source code. It's meant for engine developers to test and evaluation contributions automatically.

It's better to have the same system on all actively developed branches, especially if the new system is much faster than the old one.

And eventually the access tokens for Travis and AppVeyor might be revoked if we don't want to use them anymore.

> Shouldn't this also be changed for 3.1 and 3.0 if this is changed for 3.2?

In theory yes, but those branches are not actively developed so it's not a priority for me.
